### PR TITLE
DAOS-13814 csum: Copy iod size for checksums

### PR DESCRIPTION
**Pull request by ryon-jensen**
The iod size is required for checksums while they are being gathered on an object fetch. 
In some situations the iods structure that the checksums are using is different
than the iods structure that is used during a vos object fetch. This change copies the iod sizes
if needed.

Features: checksum
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by NiuYawei — DISMISSED** [review]

File: src/object/srv_obj.c [hunk 1]
```diff
@@ -1585,19 +1555,34 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
...
-		rc = obj_fetch_csum_init(ioc->ioc_coc, orw, orwo);
+		/**
+		 * Allocate memory for the csum structures.
+		 * This memory and information will be used by VOS to put the checksums
+		 * in as it fetches the data's metadata from the btree/evtree.
+		 *
+		 * The memory will be freed in obj_rw_reply
+		 */
+		rc = daos_csummer_alloc_iods_csums(ioc->ioc_coc->sc_csummer, iods, iods_nr, false,
+						   NULL, &orwo->orw_iod_csums.ca_arrays);
```
Comment (liuxuezhao) [inline src/object/srv_obj.c:1566]: using iods/iods_nr here seems incorrect to me.
One thing is that for the case of one IO request (update or fetch) with multiple akeys/iods, if one akey/iod targets to multiple shards, and another akey/iod target for less shards or single shards, then the some akeys/iods possibly be removed - RPC's orw->orw_iod_array with all akeys/iods, but obj_get_iods_offs() possibly return less akeys/iods (the returned nr of obj_get_iods_offs possibly less than iod_array->oia_iod_nr), and obj_get_iods_offs() returns the "skips" bitmap to indicate which akey/iod in orw->orw_iod_array is removed/skipped.

But the "orwo->orw_iod_csums.ca_arrays" should include all the csum for all items in original RPC's orw->orw_iod_array so client can check with its original RPC request.
So if change like this way here, seems incorrect to me.

But I am not sure what caused original csum issue this patch try to address.

File: src/object/srv_obj.c [hunk 2]
```diff
@@ -1585,19 +1555,34 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
...
 
 		if (ioc->ioc_coc->sc_props.dcp_csum_enabled) {
-			rc = csum_add2iods(ioh, orw->orw_iod_array.oia_iods,
-					   orw->orw_iod_array.oia_iod_nr,
-					   skips, ioc->ioc_coc->sc_csummer,
-					   orwo->orw_iod_csums.ca_arrays,
-					   orw->orw_oid, &orw->orw_dkey);
+			rc = csum_add2iods(ioh, iods, iods_nr, skips, ioc->ioc_coc->sc_csummer,
+					   orwo->orw_iod_csums.ca_arrays, orw->orw_oid,
+					   &orw->orw_dkey);
```
Comment (liuxuezhao) [inline src/object/srv_obj.c:1585]: ditto, the original code was try to use RPC's orw->orw_iod_array.oia_iods that include all original iods, but use the skip bitmap to skip some iods.

File: src/object/srv_obj.c (hunk 1 above)
Comment (ryon-jensen) [reply src/object/srv_obj.c:1566]: I see, thanks!
The issue I'm trying to solve is when there are multiple iods/akeys in a single IO request and each iod's size is DAOS_REC_ANY. The piece that's missing for checksums when iods != orw->orw_iod_array.oia_iods is the iod's size. A side effect of the vos_fetch_update call in obj_local_rw_internal is that it sets the iods' sizes.
I guess I can just copy the size from iods to orw->orw_iod_array.oia_iods. I'll work on it some more.

**Comment by ryon-jensen** [comment]
Force pushed changes because it's a completely different approach. @liuxuezhao , did I do the skip part correctly?

File: src/object/srv_obj.c [hunk 3]
```diff
@@ -1585,18 +1582,29 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
 
 	if (obj_rpc_is_fetch(rpc) && !spec_fetch &&
 	    daos_csummer_initialized(ioc->ioc_coc->sc_csummer)) {
+		if (orw->orw_iod_array.oia_iods != iods) {
+			/* Need to copy iod sizes for checksums */
+			int i, j = 0;
+
+			for (i = 0; i < orw->orw_iod_array.oia_iod_nr; i++) {
+				if (skips != NULL && isset(skips, i))
```
Comment (liuxuezhao) [inline src/object/srv_obj.c:1590]: do we need to set orw->orw_iod_array.oia_iods[i].iod_size = 0 for this skip case?
we do zero it for the skip case in obj_set_reply_sizes()

**Review by liuxuezhao — DISMISSED** [review]
LGTM, but there are a few test failures.

**Comment by ryon-jensen** [comment]
Test failure is: DAOS-13931

**Review by NiuYawei — APPROVED** [review]

File: src/common/checksum.c [hunk 4]
```diff
@@ -1156,6 +1155,7 @@ ci_insert(struct dcs_csum_info *dcb, int idx, uint8_t *csum_buf, size_t len)
 {
 	uint8_t *to_update;
 
+	D_ASSERT(dcb != NULL);
```
Comment (mchaarawi) [inline src/common/checksum.c:1158]: do we know if this assertion can be true? can this be triggered by any user input (corrupt or not) or in other circumstances?
is there no recovery beyond this point to assert the engine?

Comment (ryon-jensen) [reply src/common/checksum.c:1158]: I don't think it can, but you make a good point. I'll change it to log an error.

File: src/object/srv_obj.c [hunk 5]
```diff
@@ -1585,18 +1582,31 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
...
+
+			for (i = 0, j = 0; i < orw->orw_iod_array.oia_iod_nr; i++) {
+				if (skips != NULL && isset(skips, i)) {
+					orw->orw_iod_array.oia_iods[i].iod_size = 0;
+					continue;
+				}
+				orw->orw_iod_array.oia_iods[i].iod_size = iods[j].iod_size;
+				j++;
+			}
+		}
```
Comment (NiuYawei) [inline src/object/srv_obj.c:1597]: Is this a EC specific fix? @liuxuezhao , could you take a look?

**Review by liuxuezhao — APPROVED** [review]

File: src/object/srv_obj.c (hunk 5 above)
Comment (liuxuezhao) [reply src/object/srv_obj.c:1597]: yes, I think "if (orw->orw_iod_array.oia_iods != iods)" is only possible for EC obj.

**Comment by ryon-jensen** [comment]
pinging reviewers :-)

**Pull request merged**